cli/replicaset/cmd/upgrade.go

```diff
@@ -116,33 +118,53 @@ func getInstanceConnector(instance replicaset.Instance) (connector.Connector, er
 	})
 
 	if err != nil {
-		return nil, fmt.Errorf("instance %s failed to connect via UNIX socket "+
-			": %w", fullInstanceName, err)
+		fErr := err
+		conn, err = connector.Connect(connector.ConnectOpts{
+			Network:  connOpts.Network,
+			Address:  instance.URI,
+			Username: connOpts.Username,
+			Password: connOpts.Password,
+			Ssl:      connOpts.Ssl,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("instance %s failed to connect via UNIX socket "+
+				"and uri: %w %w", fullInstanceName, err, fErr)
+		}
 	}
 	return conn, nil
 }
 
-func collectRWROInfo(replset replicaset.Replicaset) (*instanceMeta, []instanceMeta,
+func collectRWROInfo(replset replicaset.Replicaset,
+	connOpts connector.ConnectOpts) (*instanceMeta, []instanceMeta,
 	error) {
 	var master *instanceMeta = nil
 	var replicas []instanceMeta
 	for _, instance := range replset.Instances {
 		run := instance.InstanceCtx
 		fullInstanceName := running.GetAppInstanceName(run)
-		conn, err := getInstanceConnector(instance)
+		conn, err := getInstanceConnector(instance, connOpts)
 
 		if err != nil {
 			return nil, nil, err
 		}
 
+		isRW := false
+
 		if instance.Mode == replicaset.ModeUnknown {
-			closeConnectors(master, replicas)
-			return nil, nil, fmt.Errorf(
-				"can't determine RO/RW mode on instance: %s", fullInstanceName)
+			// Discovery cannot determine all instance modes of the remote replica set.
```

Comment:
In such cases, we need to update the discovery code rather than create workrounds.